sdks/python/apache_beam/runners/portability/fn_api_runner/watermark_manager.py

```diff
@@ -0,0 +1,206 @@
+from apache_beam.runners.worker import bundle_processor
+from apache_beam.utils import proto_utils
+from apache_beam.utils import timestamp
+
+
+class WatermarkManager(object):
+  """Manages the watermarks of a pipeline's stages.
+    It works by constructing an internal graph representation of the pipeline,
+    and keeping track of dependencies."""
+  class WatermarkNode(object):
+    def __init__(self, name):
+      self.name = name
+
+  class PCollectionNode(WatermarkNode):
+    def __init__(self, name):
+      super(WatermarkManager.PCollectionNode, self).__init__(name)
+      self._watermark = timestamp.MIN_TIMESTAMP
+      self.producers = set()
+
+    def __str__(self):
+      return 'PCollectionNode<producers=[%s]' % ([i for i in self.producers])
+
+    def set_watermark(self, wm):
+      # print('setting watermark for %s to %s' % (self, wm))
+      self._watermark = min(self.upstream_watermark(), wm)
+
+    def upstream_watermark(self):
+      if self.producers:
+        return min(p.output_watermark() for p in self.producers)
+      else:
+        return timestamp.MAX_TIMESTAMP
+
+    def watermark(self):
+      if self._watermark:
+        return self._watermark
+      else:
+        return self.upstream_watermark()
+
+  class StageNode(WatermarkNode):
+    def __init__(self, name):
+      super(WatermarkManager.StageNode, self).__init__(name)
+      # We keep separate inputs and side inputs because side inputs
+      # should hold back a stage's input watermark, to hold back execution
+      # for that stage; but they should not be considered when calculating
+      # the output watermark of the stage, because only the main input
+      # can actually advance that watermark.
+      self.inputs = set()
+      self.side_inputs = set()
+
+    def __str__(self):
+      return 'StageNode<inputs=[%s],side_inputs=[%s]' % (
+          [i.name for i in self.inputs], [i.name for i in self.side_inputs])
+
+    def set_watermark(self, wm):
+      raise NotImplementedError('Stages do not have a watermark')
+
+    def output_watermark(self):
+      w = min(i.watermark() for i in self.inputs)
+      return w
+
+    def input_watermark(self):
+      w = min(i.upstream_watermark() for i in self.inputs)
+
+      if self.side_inputs:
+        w = min(w, min(i.upstream_watermark() for i in self.side_inputs))
+      return w
+
+  def __init__(self, stages):
+    # type: (List[translations.Stage]) -> None
+    self._watermarks_by_name = {}
+    for s in stages:
+      stage_name = s.name
+      stage_node = WatermarkManager.StageNode(stage_name)
+      self._watermarks_by_name[stage_name] = stage_node
+
+      # 1. Get stage inputs, create nodes for them, add to _watermarks_by_name,
+      #    and add as inputs to stage node.
+      for transform in s.transforms:
+        if transform.spec.urn == bundle_processor.DATA_INPUT_URN:
+          buffer_id = transform.spec.payload
+          if buffer_id == translations.IMPULSE_BUFFER:
```

Comment:
Should we change `IMPULSE_BUFFER` to conform to the spec of all other buffer ids rather than branching on it? (Or at least push this branch into split_buffer_id)?